### PR TITLE
executor: inline projection for hash join

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

Hash Join of #14428 

### What is changed and how it works?

Implement inline projection for Joiner, and enable it in HashJoin.

benchmark data:
```
[sunrunaway:~/tidb/executor]$ go test -run=XX -bench="BenchmarkHashJoinInlineProjection" -count 5                                   (inlineprojection-hashjoin-exec✱) 
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkHashJoinInlineProjection/InlineProjection:ON-12         	      13	  82913226 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:ON-12         	      14	  82013976 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:ON-12         	      14	  82275721 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:ON-12         	      14	  82140498 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:ON-12         	      14	  82043427 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:OFF-12        	       7	 161024914 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:OFF-12        	       7	 160023113 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:OFF-12        	       7	 162132958 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:OFF-12        	       7	 160637265 ns/op
BenchmarkHashJoinInlineProjection/InlineProjection:OFF-12        	       7	 160055525 ns/op
PASS
ok  	github.com/pingcap/tidb/executor	38.822s
```

About 100% faster for joining in memory when the SQL is like `select t1.intb from t1, t2 where t1.inta = t2.inta`

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -128,6 +128,15 @@ func (e *baseExecutor) Schema() *expression.Schema {
 	return e.schema
 }
 
+// inlineProjection supports to prune unneeded columns for outputs of a executor.
+func (e *baseExecutor) inlineProjection() (childrenUsedSchema [][]bool) {
```
Comment (XuHuaiyu) [inline executor/executor.go]: I think `inlineProjection` is not a good name for the functionality of this function. 
Can we rename it to `markUnprunedChildCol` and rename `childrenUsedSchema ` to `unprunedFlags`?

File: executor/joiner.go [hunk 2]
```diff
@@ -162,6 +196,12 @@ type baseJoiner struct {
 	selected     []bool
 	isNull       []bool
 	maxChunkSize int
+
+	// NOTE:
+	// 1. every columns are used if lUsed/rUsed is nil.
+	// 2. no columns are used if lUsed/rUsed is not nil but the size of lUsed/rUsed is 0.
+	lUsed, rUsed                   []int
```
Comment (XuHuaiyu) [inline executor/joiner.go]: rename `lUsed` and `rUsed` to `lUnpruned` and `rUnpruned`?

// 1. No column will be pruned if both lUnpruned and rUnpruned are nil.
// 2. All the columns will be pruned if lUnpruned and rUnpruned are not nil and the \`len\` of lUnpruned and rUnpruned are 0.

lUnpruned, rUnpruned

File: expression/schema.go [hunk 3]
```diff
@@ -179,6 +179,16 @@ func MergeSchema(lSchema, rSchema *Schema) *Schema {
 	return ret
 }
 
+// GetUsedList shows each schema is whether used in usedCols.
+func GetUsedList(usedCols []*Column, schema *Schema) []bool {
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply expression/schema.go]: It's copied from rule_column_pruning.go and also intended to be used there(in #14683)

File: executor/joiner.go (hunk 2 above)
Comment (SunRunAway) [reply executor/joiner.go]: I think `used` is better for human understading. It's an affirmative voice, and `unpruned` is a negetive voice which makes people think.

File: executor/executor.go (hunk 1 above)
Comment (SunRunAway) [reply executor/executor.go]: I think `used` is better for human understading. It's an affirmative voice, and `unpruned` is a negetive voice which makes people think.
And all the voice is affirmative in `rule_column_pruning.go`.

File: executor/joiner.go [hunk 4]
```diff
@@ -99,47 +107,73 @@ type joiner interface {
...
-		return &leftOuterJoiner{base}
-	case plannercore.RightOuterJoin:
-		base.chk = chunk.NewChunkWithCapacity(colTypes, ctx.GetSessionVars().MaxChunkSize)
-		return &rightOuterJoiner{base}
-	case plannercore.InnerJoin:
-		base.chk = chunk.NewChunkWithCapacity(colTypes, ctx.GetSessionVars().MaxChunkSize)
-		return &innerJoiner{base}
+	case plannercore.LeftOuterJoin, plannercore.RightOuterJoin, plannercore.InnerJoin:
+		base.chk = chunk.NewChunkWithCapacity(shallowRowType, ctx.GetSessionVars().MaxChunkSize)
+		// if conditions is not empty, we should do projection after filtering.
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/joiner.go]: 1. should -> must
2. 👍

File: expression/schema.go [hunk 5]
```diff
@@ -179,6 +179,16 @@ func MergeSchema(lSchema, rSchema *Schema) *Schema {
 	return ret
 }
 
+// GetUsedList shows each schema is whether used in usedCols.
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply expression/schema.go]: done.

File: util/chunk/chunk.go [hunk 6]
```diff
@@ -112,6 +112,22 @@ func renewColumns(oldCol []*Column, cap int) []*Column {
 	return columns
 }
 
+// renewWithCapacity creates a new Chunk based on an existing Chunk
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply util/chunk/chunk.go]: done

File: util/chunk/chunk.go [hunk 7]
```diff
@@ -167,6 +183,16 @@ func (c *Chunk) IsFull() bool {
 	return c.NumRows() >= c.requiredRows
 }
 
+// Prune return a new Chunk which prunes Columns which's index is not in usedColIdxs.
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply util/chunk/chunk.go]: done

File: executor/executor.go (hunk 1 above)
Comment (SunRunAway) [reply executor/executor.go]: And I think `inlineProjection` is better here, to describe this feature as a proper noun

File: executor/joiner.go [hunk 8]
```diff
@@ -162,6 +196,12 @@ type baseJoiner struct {
 	selected     []bool
 	isNull       []bool
 	maxChunkSize int
+
+	// NOTE:
+	// 1. every columns are used if lUsed/rUsed is nil.
+	// 2. no columns are used if lUsed/rUsed is not nil but the size of lUsed/rUsed is 0.
+	lUsed, rUsed                   []int
+	lUsedForFilter, rUsedForFilter []int
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/joiner.go]: done

File: executor/joiner.go [hunk 9]
```diff
@@ -99,47 +107,73 @@ type joiner interface {
...
+				base.lUsed = append(base.lUsed, i)
+			}
+		}
+		base.rUsed = make([]int, 0, len(projection[1])) // make it non-nil
+		for i, used := range projection[1] {
+			if used {
+				base.rUsed = append(base.rUsed, i)
+			}
+		}
+		// logutil.BgLogger().Info("InlineProjection",
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/joiner.go]: done, changed to `Debug`

File: executor/joiner.go [hunk 10]
```diff
@@ -99,47 +107,73 @@ type joiner interface {
...
+	}
 	if joinType == plannercore.LeftOuterJoin || joinType == plannercore.RightOuterJoin {
 		innerColTypes := lhsColTypes
 		if !outerIsRight {
 			innerColTypes = rhsColTypes
 		}
 		base.initDefaultInner(innerColTypes, defaultInner)
 	}
+	// shallowRowType may be different with the output columns because output columns may
+	// be inline projected, while shallow row should not be because each column may need
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/joiner.go]: done

File: executor/executor.go (hunk 1 above)
Comment (XuHuaiyu) [reply executor/executor.go]: But actually this func does not prune anything, `inlineProjection` makes the readers confused.

Comment (SunRunAway) [reply executor/executor.go]: Updated.

**Comment by SunRunAway** [comment]
@XuHuaiyu @qw4990 PTAL

File: util/chunk/chunk.go [hunk 11]
```diff
@@ -167,6 +183,17 @@ func (c *Chunk) IsFull() bool {
 	return c.NumRows() >= c.requiredRows
 }
 
+// Prune creates a new Chunk according to `c` and prunes the columns
+// whose index is not in `usedColIdxs`
+func (c *Chunk) Prune(usedColIdxs []int) *Chunk {
+	chk := renewEmpty(c)
```
Comment (qw4990) [inline util/chunk/chunk.go]: Creating a new `Chunk` whose `columns` reference to the original `Chunk` seems dangerous.
How about letting the original `Chunk` prune itself and returning nothing?
```
func (c *Chunk) Prune(usedColidxs []int)
```

Comment (SunRunAway) [reply util/chunk/chunk.go]: Sometimes the original `Chunk` should not be changed. Like in `joiner`, the original `Chunk` might be `baseJoiner.chk` which is reused in entire session and cannot be changed.

**Pull request merged**